crates/oci-distribution/src/lib.rs

```diff
@@ -0,0 +1,174 @@
+        oauth.set_client_secret(secret.unwrap_or_else(|| "".to_owned()));
+
+        let token = oauth
+            .exchange_client_credentials()
+            .with_client(&cli)
+            .execute::<StandardToken>()
+            .await?;
+
+        self.token = Some(token);
+
+        Ok(())
+    }
+
+    pub async fn pull_manifest(&self, image_name: String) -> OciResult<String> {
+        // We unwrap right now because this try_from literally cannot fail.
+        let reference = Reference::try_from(image_name).unwrap();
+        let url = reference.to_v2_manifest_url();
+        let res = reqwest::get(&url).await?;
+        let status = res.status();
+        if res.status().is_client_error() {
+            // According to the OCI spec, we should see an error in the message body.
+            let err = res.json::<OciEnvelope>().await?;
+            // FIXME: This should not have to wrap the error.
+            return Err(anyhow::format_err!("{}", err.errors[0]));
+        } else if status.is_server_error() {
+            return Err(anyhow::format_err!("Server error at {}", url));
+        }
+        let text = res.text().await?;
+        Ok(text)
+    }
+}
+
+#[derive(Clone)]
+struct BearerChallenge {
+    pub realm: Option<String>,
+    pub service: Option<String>,
+    pub scope: Option<String>,
+}
+
+impl Challenge for BearerChallenge {
+    fn challenge_name() -> &'static str {
+        "Bearer"
+    }
+    fn from_raw(raw: RawChallenge) -> Option<Self> {
+        match raw {
+            RawChallenge::Token68(_) => None,
+            RawChallenge::Fields(mut map) => Some(BearerChallenge {
+                // NB (mpb): This follows the `remove` pattern from the existing
+                // implementations. Not sure why they do this, though.
+                realm: map.remove("realm"),
+                scope: map.remove("scope"),
+                service: map.remove("service"),
+            }),
+        }
+    }
+    fn into_raw(self) -> RawChallenge {
+        let mut map = ChallengeFields::new();
+        self.realm
+            .and_then(|realm| map.insert_static_quoting("realm", realm));
+        self.scope
+            .and_then(|item| map.insert_static_quoting("scope", item));
+        self.service
+            .and_then(|item| map.insert_static_quoting("service", item));
+        RawChallenge::Fields(map)
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    #[tokio::test]
+    async fn test_version() {
+        let c = Client::default();
+        let ver = c
+            .version("webassembly.azurecr.io".to_owned())
+            .await
+            .expect("result from version request");
+        assert_eq!("registry/2.0".to_owned(), ver);
+    }
+
+    #[tokio::test]
+    async fn test_auth() {
+        let mut c = Client::default();
+        c.auth("webassembly.azurecr.io".to_owned(), None)
+            .await
+            .expect("result from version request");
```

Comment:
nit: _from `auth` request_
This still seems to fail though:

```
running 1 test
thread 'test::test_auth' panicked at 'called `Option::unwrap()` on a `None` value', /rustc/f3e1a954d2ead4e2fc197c7da7d71e6c61bad196/src/libcore/macros/mod.rs:15:40
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
test test::test_auth ... FAILED

failures:

failures:
    test::test_auth
```

Reply:
Fixed